src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseHeaderMap.cs

```diff
@@ -4,108 +4,104 @@
 using System.Collections;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
 using System.Formats.Cbor;
-using System.Runtime.Versioning;
 
 namespace System.Security.Cryptography.Cose
 {
-    public sealed class CoseHeaderMap : IEnumerable<(CoseHeaderLabel Label, ReadOnlyMemory<byte> EncodedValue)>
+    public sealed class CoseHeaderMap : IDictionary<CoseHeaderLabel, CoseHeaderValue>, IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>
     {
         private static readonly CoseHeaderMap s_emptyMap = new CoseHeaderMap(isReadOnly: true);
 
         public bool IsReadOnly { get; internal set; }
 
-        private readonly Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>> _headerParameters = new Dictionary<CoseHeaderLabel, ReadOnlyMemory<byte>>();
+        public ICollection<CoseHeaderLabel> Keys => _headerParameters.Keys;
 
-        public CoseHeaderMap() : this (isReadOnly: false) { }
+        public ICollection<CoseHeaderValue> Values => _headerParameters.Values;
 
-        internal CoseHeaderMap(bool isReadOnly)
-        {
-            IsReadOnly = isReadOnly;
-        }
+        public int Count => _headerParameters.Count;
 
-        public bool TryGetEncodedValue(CoseHeaderLabel label, out ReadOnlyMemory<byte> encodedValue)
-            => _headerParameters.TryGetValue(label, out encodedValue);
+        IEnumerable<CoseHeaderLabel> IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>.Keys => _headerParameters.Keys;
 
-        public ReadOnlyMemory<byte> GetEncodedValue(CoseHeaderLabel label)
+        IEnumerable<CoseHeaderValue> IReadOnlyDictionary<CoseHeaderLabel, CoseHeaderValue>.Values => _headerParameters.Values;
+
+        public CoseHeaderValue this[CoseHeaderLabel key]
         {
-            if (TryGetEncodedValue(label, out ReadOnlyMemory<byte> encodedValue))
+            get => _headerParameters[key];
+            set
             {
-                return encodedValue;
+                ValidateIsReadOnly();
+                ValidateInsertion(key, value);
+                _headerParameters[key] = value;
             }
-
-            throw new InvalidOperationException(SR.Format(SR.CoseHeaderMapLabelDoeNotExist, label.LabelName));
         }
 
-        public int GetValueAsInt32(CoseHeaderLabel label)
-        {
-            var reader = new CborReader(GetEncodedValue(label));
-            int retVal = reader.ReadInt32();
-            Debug.Assert(reader.BytesRemaining == 0);
-            return retVal;
-        }
+        private readonly Dictionary<CoseHeaderLabel, CoseHeaderValue> _headerParameters = new Dictionary<CoseHeaderLabel, CoseHeaderValue>();
```

Comment:
The field should be declared back where it was (above these properties and indexers)